### PR TITLE
Add redis lock feature

### PR DESCRIPTION
**Pull request by Bibob7**
## Description

Based on the discussion in my previous PR: [#1031](https://github.com/oauth2-proxy/oauth2-proxy/pull/1031), I tried to implement the proposal.

## Motivation and Context

If parallel requests hitting the oauth2-proxy with an expired session, oauth2-proxy will perform multiple requests to the token endpoint to refresh it.
Because of refresh token rotation, a refresh token can only be used once, so that the second request to the token endpoint will fail.
In this case, it is possible to repeat the retrieval of the session from the persistent storage, as the successful request must have already updated the session in the persistent storage.

Issue: #1006

## How Has This Been Tested?

I used the implemented "Lock" and "Release" method from SessionState before and after refreshing the session. The Redis lock feature worked properly.

I also added tests in `pkg/sessions/tests/session_store_tests.go`

## Checklist:

- [X] My change requires a change to the documentation or CHANGELOG.
- [X] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: pkg/middleware/stored_session.go
Comment (NickMeves) [inline pkg/middleware/stored_session.go:110]: I think somewhere around here you need to handle nodes that don't receive the lock (e.g. NodeA gets the lock and performs the refresh, NodeB is too late, ticket is locked).

I think NodeB needs to block here until the lock is released. Then instead of getting the lock and trying to refresh itself, it pulls the session from storage (and uses it assuming NodeA just refreshed it).

File: pkg/sessions/redis/redis_store.go
Comment (NickMeves) [inline pkg/sessions/redis/redis_store.go:65]: This potentially feels too short.

We don't have data, but we'd ideally want P90 or P99 response times to IdP refresh endpoints.  I have no idea what that would be though... 🤔 

1s feels right (it normally releases earlier manually when the operation was faster, right?)

File: pkg/sessions/redis/client.go
Comment (NickMeves) [inline pkg/sessions/redis/client.go:8]: Good find! I looked at their source code -- the Lua they used to keep the lock/release in a single transaction looks the same as what I did 3-4 years ago.  We must've read the same design article 🤣

File: pkg/sessions/redis/client.go
Comment (NickMeves) [inline pkg/sessions/redis/client.go:33]: What's the rationale for tracking locks ourselves vs relying on underlying errors returned by `locker.Obtain` if a lock is already taken?

Do we run the risk of drift if we try to manage assignment & expiration in 2 different spots (in our code, & redis data store)? I understand the benefits of shaving roundtrips to Redis, so something to weight pros and cons of.

File: pkg/sessions/redis/client.go
Comment (NickMeves) [inline pkg/sessions/redis/client.go:52]: It looks like `redislock` doesn't have a way to `Peek` at whether a lock is already taken without trying to obtain it ourselves? I really only see `Obtain` exposed.

That functionality would be nice for us to block and wait for a lock to be done without trying to take it (so we can then pull a fresh session from the store).

File: pkg/sessions/redis/client.go
Comment (NickMeves) [reply pkg/sessions/redis/client.go:33]: While this also tracks all locks potentially acquired in a single process space -- this would fall apart in horizonatally scaled deployments right?

File: pkg/sessions/redis/client.go
Comment (NickMeves) [reply pkg/sessions/redis/client.go:52]: I want to avoid a long queue of requests waiting for the session refresh lock to be done - only to have to all lock/unlock themselves to pull the fresh session.

File: pkg/sessions/redis/client.go
Comment (Bibob7) [reply pkg/sessions/redis/client.go:33]: Based on the redislock implementation, I need to know the lock to release it. I simply call `lock.Release()`. Thats why I'm persisting it here.

File: pkg/sessions/redis/client.go
Comment (Bibob7) [reply pkg/sessions/redis/client.go:52]: You are right, unfortunately, they are not supporting that right now.

Comment (Bibob7) [reply pkg/sessions/redis/client.go:52]: I solved it with just doing retries with Load, if LoadWithLock failed once. What do you think. This is not yet tested, just an idea.

File: pkg/sessions/redis/client.go
Comment (NickMeves) [reply pkg/sessions/redis/client.go:33]: So I think this might have issues since maps aren't safe for concurrency without using a mutex or channels. Plus it feels weird to manage locks on the client.

What about this design:

We have our own Lock interface (calling out `Obtain` / `Release`, whatever else we need). Then we store the Lock in whatever `SessionState` we pop out of our `SessionStore` (so our store will still `Save`, `Load`, `Clear`).

So the Redis store will attach a `Lock` implementation that is `redislock` aware to the created `SessionState`, Cookie will attach its own noop Lock.

Then we can fully manage and control the lock from the SessionState if they ever have an operation that needs a lock (e.g. refresh) - that feels more natural to me than trying to track all existing locks from the client.

Comment (NickMeves) [reply pkg/sessions/redis/client.go:33]: This design would go better once we refactored our `RefreshSession` implementation to allow refreshing before the Access Token has expired.

Then 1 request `SessionState` could grab the lock for refreshing, while the others peek (we can check the redis key) and skip refreshing and carry on proxying like normal if the refreshing lock is claimed by another request.

Otherwise, if you can't refresh until the access token is expired, I take it you want to have all requests wait until the session is refreshed? Or can we be greedy and have them be slightly stale and the follow-up requests that aren't concurrent with the one that grabbed the refresh lock will be able to pull up the refreshed session 🤔

Comment (Bibob7) [reply pkg/sessions/redis/client.go:33]: > We have our own Lock interface (calling out Obtain / Release, whatever else we need). Then we store the Lock in
> whatever SessionState we pop out of our SessionStore (so our store will still Save, Load, Clear).
> 
> So the Redis store will attach a Lock implementation that is redislock aware to the created SessionState, Cookie will 
> attach its own noop Lock.
> 
> Then we can fully manage and control the lock from the SessionState if they ever have an operation that needs a lock > (e.g. refresh) - that feels more natural to me than trying to track all existing locks from the client.

Yes, I like the idea. I will try to change it like that.

I also like the idea of refreshing when the access token is still valid for X amount of time. In this case we can actually just skip refreshing. However, this would increase the necessary changes here in the PR again. What do you suggest how to proceed here?

What do you think about my further adjustments regarding the isRefreshSessionNeededWithProvider method?

Comment (NickMeves) [reply pkg/sessions/redis/client.go:33]: Let's solely handle locking in this PR -- its a complex topic and it will help @JoelSpeed and I fully understand the nuances of just that as we wrap our brains around it.

Early refreshing is another complex refactor with its share of pitfalls. We are tracking it here: https://github.com/oauth2-proxy/oauth2-proxy/issues/856

**Comment by JoelSpeed** [comment]
I've not reviewed this just yet, super busy at the moment. But I noticed some conversation about storing a lock or lock key. Have we considered using the request scope to store this? It is unique to each request so should be able to be used to store whatever we need to track across the lock/unlock cycle as I understand it

**Comment by gysel** [comment]
@Bibob7, let me know when the code is ready to be tested. We also have issues with concurrent refreshes.

File: pkg/sessions/redis/client.go
Comment (JoelSpeed) [reply pkg/sessions/redis/client.go:41]: > doesn't have a way to Peek at whether a lock is already taken

Is this something we could implement on top of it possibly?

> I want to avoid a long queue of requests waiting for the session refresh lock to be done - only to have to all lock/unlock themselves to pull the fresh session.

Lock/unlock should be pretty quick once we know that the session has been revalidated right? What I don't want to do is overcomplicate the original implementation by having strict requirements that may not be 100% necessary. This PR is quite large, I wonder if we can break it into steps and get a minimal implementation that may not be polished, and iterate to make it more efficient?

Comment (NickMeves) [reply pkg/sessions/redis/client.go:41]: I'm gonna try to spike out an implementation of the proactive refreshing that is tied to this -- that being separate might make us able to fully focus on the locking functionality in this PR.

Comment (Bibob7) [reply pkg/sessions/redis/client.go:41]: @NickMeves: Great, let me know, when this is done and I can continue here.

Comment (NickMeves) [reply pkg/sessions/redis/client.go:41]: Draft is up here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1086

Once I get some feedback on the design choices from @JoelSpeed I'll wrap it up. It's admittedly raw since I did it from my couch while watching TV 😅

**Comment by Bibob7** [comment]
I reduced everything in this PR to locking functionality. Please have a look, it it makes sense for you. Is tested it with a few adjustment in `pkg/middleware/stored_session.go ` by using the lock and release methods. It worked!

The adjustments are not part of this PR. I think we can apply that, when #1086 is done.

**Comment by NickMeves** [comment]
> I reduced everything in this PR to locking functionality. Please have a look, it it makes sense for you. Is tested it with a few adjustment in `pkg/middleware/stored_session.go ` by using the lock and release methods. It worked!
> 
> The adjustments are not part of this PR. I think we can apply that, when #1086 is done.

Thanks @Bibob7 -- We're wrapping up v7.1 at the moment, I'll give this another more thorough review after we wrap up the PRs on the critical path for v7.1's release.

File: pkg/apis/sessions/interfaces.go
Comment (JoelSpeed) [inline pkg/apis/sessions/interfaces.go:17]: We should probably use the errors package for this
```suggestion
var ErrLockNotObtained = errors.New("lock: not obtained")
```

File: pkg/sessions/persistence/manager.go
Comment (JoelSpeed) [inline pkg/sessions/persistence/manager.go:69]: No need to wrap this in another anonymous function is there? I think this should work
```suggestion
		m.Store.Lock,
	)
```

File: pkg/sessions/persistence/ticket.go
Comment (JoelSpeed) [inline pkg/sessions/persistence/ticket.go:35]: Minor observation, not necessarily a problem, I wonder if the name of this implies it locks when it's called? Is there a more explicit name we can use to make sure that's not confusing? Is it just me? (This isn't a blocker just a thought for conversation)

File: pkg/sessions/redis/lock.go
Comment (JoelSpeed) [inline pkg/sessions/redis/lock.go:69]: Don't lock keys normally use a suffix for locking? We should double check if there's a convention or good reason for doing this

File: pkg/sessions/redis/lock.go
Comment (JoelSpeed) [inline pkg/sessions/redis/lock.go:62]: Should we have a constant in the sessions for this kind of error? Something like `ErrNotLocked`?

File: pkg/sessions/redis/lock.go
Comment (JoelSpeed) [inline pkg/sessions/redis/lock.go:32]: Just in case there's any wrapping going on here, is it worth using `errors.Is()`?

File: pkg/sessions/tests/mock_store.go
Comment (JoelSpeed) [inline pkg/sessions/tests/mock_store.go:67]: This doesn't seem to be the opposite of release lock? Shouldn't there be a MockLock in use here? Or do we somehow need to do something with the lock cache?

**Review by JoelSpeed — COMMENTED** [review]
This is really interesting and I think pretty close to being ready to go! I'd love to see some more comments throughout (especially on things like interfaces) and if possible some mock locking and tests around this area.

Also wondering if we should add a set of lockable tests to the session store test suite, ie, can we define some tests that a session store implementing a lock should be able to pass?

File: pkg/sessions/persistence/ticket.go
Comment (Bibob7) [reply pkg/sessions/persistence/ticket.go]: I think `initLockFunc` is a better name here.

**Comment by salmanisd** [comment]
Hi guys... any update when this PR can be merged? Currently running into the same invalid refresh tokeb problem.

**Comment by Bibob7** [comment]
@JoelSpeed & @NickMeves: Are there still changes that need to be done? I would like to merge this PR, if possible. 😊

**Comment by NickMeves** [comment]
> @JoelSpeed & @NickMeves: Are there still changes that need to be done? I would like to merge this PR, if possible. 😊

I'll need to defer to @JoelSpeed on getting this one across the finish line. I just started at a new job so I won't have much time for any deep OAuth2 Proxy work or code reviews for a few weeks.

Saving all my mental stamina for all the new stuff I need to learn during onboarding 😅

File: CHANGELOG.md
Comment (JoelSpeed) [inline CHANGELOG.md:30]: This needs to be moved to the latest changes since, as we have had a new release since this was last rebased

File: pkg/apis/sessions/interfaces.go
Comment (JoelSpeed) [inline pkg/apis/sessions/interfaces.go:26]: Would be good to have comments on the functions in the interface so we know what they do/expected behaviour etc.

This will aid future lock implementors to understand the interface

File: pkg/sessions/redis/lock.go
Comment (JoelSpeed) [reply pkg/sessions/redis/lock.go:84]: This question is still outstanding

**Review by JoelSpeed — COMMENTED** [review]
This is pretty good to go IMO. Would like to see some documentation on the interface so we know what the expectations are for future implementors. And there's one comment still outstanding that I had from march around whether we should use a suffic for the lock, this seems to be more common from what I've seent.

Otherwise looks great! Appreciate all the work you've put into this @Bibob7

File: pkg/sessions/redis/lock.go
Comment (NickMeves) [inline pkg/sessions/redis/lock.go:14]: This should be a `const` right?

File: pkg/sessions/redis/lock.go
Comment (NickMeves) [inline pkg/sessions/redis/lock.go:47]: I'm curious on Refresh & Release:

If I recall, the underlying `redislock` library has lock protection be having a cryptographic nonce in the value. The locker also has that value and any modification to the lock needs to validate that the caller knows the cryptographic nonce (protects against a non-holder of the nonce from trampling by accessing the same lock key).

What is the rationale for messing with the lock keys in Redis directly without using the `locker` that gives these protections?

`Release` in particular I wonder if there's risk of trampling another lock holder's lock?

Comment (NickMeves) [reply pkg/sessions/redis/lock.go:47]: https://github.com/bsm/redislock/blob/162618b021216846fc29f90b1c58f2620131714b/redislock.go#L133

Comment (Bibob7) [reply pkg/sessions/redis/lock.go:47]: Thats' a good point. I did it to actually enable this specific case. But you are right, it should not be like that. I will update it by using the lock instance for refreshing and releasing.

**Comment by Bibob7** [comment]
What do you say, is this ready to merge? 😊 @JoelSpeed @NickMeves 
For the lock key, I also switched to suffix.

**Review by JoelSpeed — APPROVED** [review]
I think this is good to go now, if there are any further issues, let's merge them in follow ups

Thanks for your patience @Bibob7, I think we've got to a state of high quality code for the addition of this feature!

**Pull request merged**
